### PR TITLE
Input through file and pipe

### PR DESCRIPTION
**Pull request by mao3267**
## Tracking issue
Related to flyteorg/flyte#5365

## Why are the changes needed?
1. We want to input multiple parameters through a single JSON/YAML file. eg. `pyflyte run wf.py --inputs-file input.json`
2. We want to support input from pipe as well. eg. `cat input.json | pyflyte run wf.py wf`

## What changes were proposed in this pull request?
1. Add a new option `--inputs-file` to specify the input file.
2. Read from file or pipe and transform into click options
3. Add unit test (support all types, replace the input from file/pipe if there's an option specify the same parameter)

## How was this patch tested?
Currently, two unit tests are performed on this feature. `test_all_types_with_file_input` and `test_replace_file_inputs` are added in `tests/flytekit/unit/cli/pyflyte/test_run.py`

#### Missing value in custom class
`wf.py`
```
from flytekit import task, workflow
from dataclasses import dataclass
import typing
from dataclasses_json import DataClassJsonMixin


@dataclass
class myclass(DataClassJsonMixin):
    i: int
    j: typing.List[int]


@task
def t1(a: myclass) -> myclass:
    return a


@workflow
def wf(a: myclass) -> myclass:
    return t1(a=a)
```
`input.json`
```
{
    "a": {
        "j": [1, 2, 3]
    }
}
```
Output:
image

#### Missing input parameter
`wf.py`
```
from flytekit import task, workflow

@task
def t1(a: int, b: int) -> int:
    return a + b

@workflow
def wf(a: int, b: int) -> int:
    return t1(a=a, b=b)
```
`input.json`
```
{
    "a": 1,
}
```
Output:
image

#### Help message
image

More tests can be added if needed.
### Setup process
```sh
git clone https://github.com/mao3267/flytekit.git
git checkout input-through-file-and-pipe
make setup && pip install -e .
```
### Screenshots
None
## Check all the applicable boxes 

- [ ] I updated the documentation accordingly.
- [x] All new and existing tests passed. (test_run.py)
- [x] All commits are signed-off.

## Related PRs
#2583

## Docs link
ToDo

### CODE REVIEW

**Comment by kumare3** [comment]
@mao3267 can you please DCO sign-off, This is awesome would love to get it merged

File: flytekit/clis/sdk_in_container/run.py [hunk 1]
```diff
@@ -763,6 +810,87 @@ def get_command(self, ctx, name):
...
+                        inputs = json.loads(f)
+                    except json.JSONDecodeError as e:
+                        raise click.BadParameter(
+                            message=f"Could not load the inputs file. Please make sure it is a valid JSON or YAML file."
+                            f"\n json error: {e},"
+                            f"\n yaml error: {yaml_e}",
+                            param_hint="--inputs-file",
+                        )
+                new_args = []
+                for k, v in inputs.items():
```
Comment (kumare3) [inline flytekit/clis/sdk_in_container/run.py:880]: can we refactor this as this is repeated with lines 855-860. You can refactor it to separate block after the yaml has been loaded?

**Comment by mao3267** [comment]
@kumare3 I have refactored the duplicate code and signed off my commits. If any further changes are needed, please let me know!

File: flytekit/clis/sdk_in_container/run.py [hunk 2]
```diff
@@ -775,6 +819,76 @@ def get_command(self, ctx, name):
...
+                except yaml.YAMLError as e:
+                    yaml_e = e
+                    try:
+                        inputs = json.loads(f)
+                    except json.JSONDecodeError as e:
+                        raise click.BadParameter(
+                            message=f"Could not load the inputs file. Please make sure it is a valid JSON or YAML file."
+                            f"\n json error: {e},"
+                            f"\n yaml error: {yaml_e}",
+                            param_hint="--inputs-file",
```
Comment (pingsutw) [inline flytekit/clis/sdk_in_container/run.py:874]: Could we create a common function to read json/yaml file for `stdin` and `--input-file`?

File: flytekit/clis/sdk_in_container/run.py [hunk 3]
```diff
@@ -819,21 +933,34 @@ def _create_command(
 
         # Add options for each of the workflow inputs
         params = []
-        for input_name, input_type_val in loaded_entity.python_interface.inputs_with_defaults.items():
+        for (
```
Comment (pingsutw) [inline flytekit/clis/sdk_in_container/run.py:936]: why do we need the parentheses here

Comment (mao3267) [reply flytekit/clis/sdk_in_container/run.py]: The parentheses should be removed. Thanks!

File: flytekit/clis/sdk_in_container/run.py (hunk 2 above)
Comment (mao3267) [reply flytekit/clis/sdk_in_container/run.py]: @pingsutw A common function `load_inputs` is added for parsing inputs from `stdin` and `--inputs-file`. Hope it looks good to us ~

**Review by pingsutw — DISMISSED** [review]

**Review by pingsutw — APPROVED** [review]

**Comment by mao3267** [comment]
Thanks for the help from @pingsutw, @kumare3 and other reviewers! Glad to see it merged.

**Comment by kumare3** [comment]
Can we add docs for this - this is a very good and important feature for folks experimenting continuously

**Comment by pingsutw** [comment]
@mao3267 yeah, thanks for the contribution. Could you help update this [doc](https://docs.flyte.org/en/latest/api/flytekit/design/clis.html#what-is-pyflyte-run) too?

**Pull request merged**